plugin/pkg/utils/cloudprofile.go

```diff
@@ -0,0 +1,143 @@
+		if err != nil {
+			return nil, err
+		}
+		return &gardencorev1beta1.CloudProfile{Spec: namespacedCloudProfile.Status.CloudProfileSpec}, nil
+	case constants.CloudProfileReferenceKindCloudProfile:
+		return cloudProfileLister.Get(cloudProfileReference.Name)
+	}
+	return nil, fmt.Errorf("could not find referenced cloudprofile")
+}
+
+// ValidateCloudProfileChanges validates that the referenced CloudProfile does only change towards a more specific reference
+// (i.e. currently only from a CloudProfile to a descendant NamespacedCloudProfile).
+// For now, other changes are not supported (e.g. from one CloudProfile to another or from one NamespacedCloudProfile to another).
+func ValidateCloudProfileChanges(_ gardencorev1beta1listers.CloudProfileLister, namespacedCloudProfileLister gardencorev1beta1listers.NamespacedCloudProfileLister, newShoot, oldShoot *core.Shoot) error {
+	oldCloudProfileReference := BuildCloudProfileReference(oldShoot)
+	if oldCloudProfileReference == nil {
+		return nil
+	}
+	newCloudProfileReference := BuildCloudProfileReference(newShoot)
+	if equality.Semantic.DeepEqual(oldCloudProfileReference, newCloudProfileReference) {
+		return nil
+	}
+
+	newCloudProfileRoot, err := getRootCloudProfile(namespacedCloudProfileLister, newCloudProfileReference, newShoot.Namespace)
+	if err != nil {
+		return err
+	}
+	if oldCloudProfileReference.Kind == constants.CloudProfileReferenceKindCloudProfile &&
+		newCloudProfileReference.Kind == constants.CloudProfileReferenceKindNamespacedCloudProfile &&
+		equality.Semantic.DeepEqual(oldCloudProfileReference, newCloudProfileRoot) {
+		return nil
+	}
+
+	return fmt.Errorf("a Seed's CloudProfile may only be changed to a descendant NamespacedCloudProfile, other modifications are currently not supported")
+}
+
+// getRootCloudProfile determines the root CloudProfile from a CloudProfileReference containing any (Namespaced)CloudProfile
+func getRootCloudProfile(namespacedCloudProfileLister gardencorev1beta1listers.NamespacedCloudProfileLister, cloudProfile *gardencorev1beta1.CloudProfileReference, namespace string) (*gardencorev1beta1.CloudProfileReference, error) {
+	if cloudProfile == nil {
+		return nil, errors.New("unexpected nil cloudprofile to get root of")
+	}
+	switch cloudProfile.Kind {
+	case constants.CloudProfileReferenceKindCloudProfile:
+		return cloudProfile, nil
+	case constants.CloudProfileReferenceKindNamespacedCloudProfile:
+		cp, err := namespacedCloudProfileLister.NamespacedCloudProfiles(namespace).Get(cloudProfile.Name)
+		if err != nil {
+			return nil, err
+		}
+		return getRootCloudProfile(namespacedCloudProfileLister, &cp.Spec.Parent, namespace)
+	}
+	return nil, fmt.Errorf("unexpected cloudprofile kind %s", cloudProfile.Kind)
+}
+
+// BuildCloudProfileReference determines the CloudProfile of a Shoot to use
+// depending on the availability of cloudProfile and cloudProfileName.
```

Comment:
```suggestion
// BuildCloudProfileReference determines and returns the CloudProfile reference of the given shoot,
// depending on the availability of cloudProfileName and cloudProfile.
```

Reply:
was the order change of `cloudProfile` and `cloudProfileName` intentional?

Reply:
Nope 😄